Python/_warnings.c

```diff
@@ -232,14 +232,77 @@ get_warnings_attr(PyInterpreterState *interp, PyObject *attr, int try_import)
     return obj;
 }
 
+static inline void
+warnings_lock(PyInterpreterState *interp)
+{
+#ifdef Py_GIL_DISABLED
```

Comment:
This pure python impl uses mutex is both gil and nogil builds. The same should happen in C impl too I guess?

Reply:
Yeah, I think that's right.  The nogil build can have data races as well and so that code needs a lock to avoid that.